File: internal/sam/receiver.go
package sam

import (
	"context"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"net"
	"strings"
	"time"

	"github.com/go-i2p/sam3"
)

type ReceiverConfig struct {
	SAMAddr string
	Size    int64
	Seed    int64
	ID      string
	Timeout time.Duration
}

func RunReceiver(cfg ReceiverConfig, w io.Writer) error {
	ctx, cancel := context.WithTimeout(context.Background(), cfg.Timeout)
	defer cancel()

	res := ResultMsg{
		Type:    MsgResult,
		Role:    "receiver",
		Success: false,
	}
	enc := json.NewEncoder(w)

	fail := func(format string, args ...any) error {
		res.Error = fmt.Sprintf(format, args...)
		enc.Encode(res)
		return errors.New(res.Error)
	}

	s, err := sam3.NewSAM(cfg.SAMAddr)
	if err != nil {
		return fail("SAM connect: %v", err)
	}
	defer s.Close()

	keys, err := s.NewKeys()
	if err != nil {
		return fail("NewKeys: %v", err)
	}

	session, err := s.NewStreamSession(cfg.ID, keys, streamTunnelOptions())
	if err != nil {
		return fail("NewStreamSession: %v", err)
	}
	defer session.Close()

	listener, err := session.Listen()
	if err != nil {
		return fail("Listen: %v", err)
	}
	defer listener.Close()

	// Публикуем полный base64-ключ destination, а не только b32-хэш.
	// DialI2P на sender-стороне принимает i2pkeys.I2PAddr (полный ключ).
	if err := enc.Encode(ReadyMsg{Type: MsgReady, Dest: string(keys.Addr())}); err != nil {
		return err
	}

	type acceptResult struct {
		conn net.Conn
		err  error
	}

	for {
		tAccept := time.Now()
		ch := make(chan acceptResult, 1)
		go func() {
			c, e := listener.Accept()
			ch <- acceptResult{c, e}
		}()

		var conn net.Conn
		select {
		case <-ctx.Done():
			return fail("timeout waiting for connection")
		case r := <-ch:
			if r.err != nil {
				return fail("Accept: %v", r.err)
			}
			conn = r.conn
		}

		received, shaOk, firstByteAt, err := ReceivePayload(conn)
		conn.Close()
		tDone := time.Now()

		// i2pd/SAM can occasionally deliver a reset/empty accepted stream before
		// the real payload stream. Ignore it and keep accepting until timeout.
		if err != nil && received == 0 && strings.Contains(err.Error(), "failed to read size: EOF") {
			continue
		}

		res.PayloadBytes = received
		res.SHA256OK = shaOk
		if err != nil {
			return fail("ReceivePayload: %v", err)
		}

		res.Success = true
		res.FirstByteMs = float64(firstByteAt.Sub(tAccept).Milliseconds())
		res.TransferMs = float64(tDone.Sub(firstByteAt).Milliseconds())

		return enc.Encode(res)
	}
}

File: testbed/src/sam/receiver.rs
use super::{
    messages::{ReadyMsg, ResultMsg},
    wire::receive_payload,
};
use std::{
    error::Error,
    io::Write,
    thread,
    time::{Duration, Instant},
};

use super::sam3::{SamSession, SAM_TUNNEL_OPTIONS};

pub struct ReceiverConfig {
    pub sam_addr: String,
    pub size: i64,
    pub seed: i64,
    pub id: String,
    pub timeout: Duration,
}

pub fn run_receiver(cfg: ReceiverConfig, w: &mut impl Write) -> Result<(), Box<dyn Error>> {
    let deadline = Instant::now() + cfg.timeout;
    let mut res = ResultMsg::new("receiver");
    let fail =
        |res: &mut ResultMsg, w: &mut dyn Write, msg: String| -> Result<(), Box<dyn Error>> {
            res.error = msg.clone();
            serde_json::to_writer(&mut *w, res)?;
            writeln!(w)?;
            Err(msg.into())
        };

    let session = match SamSession::create_stream(&cfg.sam_addr, &cfg.id, SAM_TUNNEL_OPTIONS) {
        Ok(s) => s,
        Err(e) => return fail(&mut res, w, format!("SAM session: {e}")),
    };

    serde_json::to_writer(&mut *w, &ReadyMsg::new(session.destination.clone()))?;
    writeln!(w)?;
    w.flush()?;

    loop {
        if Instant::now() >= deadline {
            return fail(&mut res, w, "timeout waiting for connection".to_string());
        }

        let t_accept = Instant::now();
        let mut conn = match SamSession::accept_stream(&cfg.sam_addr, &cfg.id) {
            Ok(c) => c,
            Err(e) => return fail(&mut res, w, format!("Accept: {e}")),
        };

        let result = receive_payload(&mut conn);
        let t_done = Instant::now();
        drop(conn);

        match result {
            Ok((received, sha_ok, first_byte_at)) => {
                res.payload_bytes = received;
                res.sha256_ok = sha_ok;
                res.success = true;
                res.first_byte_ms = first_byte_at.duration_since(t_accept).as_millis() as f64;
                res.transfer_ms = t_done.duration_since(first_byte_at).as_millis() as f64;
                drop(session);
                serde_json::to_writer(&mut *w, &res)?;
                writeln!(w)?;
                return Ok(());
            }
            Err(e) => {
                let msg = e.to_string();
                if msg.contains("failed to read size:") && msg.contains("EOF") {
                    thread::sleep(Duration::from_millis(100));
                    continue;
                }
                res.payload_bytes = 0;
                return fail(&mut res, w, format!("ReceivePayload: {msg}"));
            }
        }
    }
}
